File: 904. Fruit Into Baskets/fruit_into_baskets.py
# T.C.->O(n)
# S.C.->O(1)
from collections import defaultdict
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        n = len(fruits)
        mp = defaultdict(int)
        i = 0
        count = 0

        for j in range(n):
            mp[fruits[j]] += 1
            while len(mp) > 2:
                mp[fruits[i]] -= 1
                if mp[fruits[i]] == 0:
                    del mp[fruits[i]]
                i += 1
            count = max(count, j - i + 1)

        return count

File: 904. Fruit Into Baskets/fruit_into_baskets.cpp
// T.C.->O(n)
// S.C.->O(1)

// Largest Window(subarray) having at most 2 unique type of fruits
class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        int n=fruits.size();
        unordered_map<int,int> mp;
        int i=0;
        int j=0;
        int count=0; //result
        while(j<n)
        {
            mp[fruits[j]]++;
            if(mp.size()<=2)
                count = max(count,j-i+1);
            else{
                mp[fruits[i]]--;
                if(mp[fruits[i]] == 0)
                    mp.erase(fruits[i]);
                i++;
            }
            j++;
        }
        return count;
    }
};
